src/server/main_service.cc

```diff
@@ -1326,6 +1325,196 @@ bool StartMultiExec(DbIndex dbid, Transaction* trans, ConnectionState::ExecInfo*
+    f(args[i]);
+
+  if (keys.bonus)
+    f(args[keys.bonus]);
+}
+
+// ExecutionSquasher allows executing a series of commands under a multi transaction
+// and squashing multiple consecutive single-shard commands into one hop whenever it's possible,
+// thus greatly decreasing the dispatch overhead for them.
+class ExecutionSquasher {
+ public:
+  static void Execute(absl::Span<StoredCmd> cmds, ConnectionContext* cntx) {
+    ExecutionSquasher{cmds, cntx}.Run();
+  }
+
+ private:
+  using ReplyChan = ::util::fibers_ext::SimpleChannel<CapturingReplyBuilder::Payload>;
+
+  // Per-shard exection info.
+  struct ShardExecInfo {
+    ShardExecInfo() : cmds{}, reply_chan{nullptr}, local_tx{nullptr} {
+    }
+
+    std::vector<StoredCmd*> cmds;  // accumulated commands
+    unique_ptr<ReplyChan> reply_chan;
+    intrusive_ptr<Transaction> local_tx;  // stub-mode tx for use inside shard
+  };
+
+  static constexpr int kChanBufferSize = 16;
+
+ private:
+  ExecutionSquasher(absl::Span<StoredCmd> cmds, ConnectionContext* cntx);
+
+  // Lazy initialize shard info
+  ShardExecInfo& PrepareShardInfo(ShardId sid);
+
+  // Return true if command was squashed.
+  bool TrySquash(StoredCmd* cmd);
+
+  // Execute separate non-squashed cmd.
+  void ExecuteStandalone(StoredCmd* cmd);
+
+  // Callback that runs on shards during squashed hop.
+  OpStatus SquashedHopCb(Transaction* parent_tx, EngineShard* es);
+
+  // Execute all currently squashed commands.
+  void ExecuteSquashed();
+
+  // Run all commands until completion.
+  void Run();
+
+ private:
+  absl::Span<StoredCmd> cmds_;  // Input range of stored commands
+  ConnectionContext* cntx_;     // Underlying context
+  const CommandId* base_cid_;   // either EVAL or EXEC, used for squashed hops
+
+  std::vector<ShardExecInfo> sharded_;
+  std::vector<ShardId> order_;  // reply order for squashed cmds
+
+  // multi modes that lock on hops (non-atomic, incremental) need keys for squashed hops.
+  // track_keys_ stores whether to populate collected_keys_
+  bool track_keys_;
+  absl::flat_hash_set<MutableSlice> collected_keys_;
+
+  absl::flat_hash_set<ShardId> tmp_shards_;
+  std::vector<MutableSlice> tmp_keylist_;
+  std::vector<uint8_t> sharded_reuse_list_;
+};
+
+ExecutionSquasher::ExecutionSquasher(absl::Span<StoredCmd> cmds, ConnectionContext* cntx)
+    : cmds_{cmds}, cntx_{cntx}, base_cid_{cntx->transaction->GetCId()} {
+  auto mode = cntx->transaction->GetMultiMode();
+  track_keys_ = (mode == Transaction::LOCK_INCREMENTAL) || (mode == Transaction::NON_ATOMIC);
+}
+
+ExecutionSquasher::ShardExecInfo& ExecutionSquasher::PrepareShardInfo(ShardId sid) {
+  if (sharded_.empty())
+    sharded_.resize(shard_set->size());
+
+  auto& sinfo = sharded_[sid];
+  if (!sinfo.local_tx)
+    sinfo.local_tx = new Transaction{cntx_->transaction};
+
+  if (!sinfo.reply_chan)
+    sinfo.reply_chan = make_unique<ReplyChan>(kChanBufferSize);
+
+  return sinfo;
+}
```

Comment:
1. Transaction manages its own lifetime as a pointer (because cb's can be stuck?), so we need it to be a pointer
2. The channel is immovable - so hide it behind a pointer
3. I use direct indexing (like tx does), so the vector as large as the shard_set, because we have no better general solution (plain array? skiplist?)

We can also implement re-use of those resources, so that if we have only a couple of shards active at a time we'll create only a few resources and re-use them.

Even though its imperfect, those steps sound like premature optimization - we first need to see in detail what overhead we introduce